libminifi/src/utils/Cron.cpp

```diff
@@ -0,0 +1,510 @@
+}
+
+template <>
+seconds getFieldType(local_seconds time_point) {
+  auto dp = floor<days>(time_point);
+  auto time = make_time(time_point-dp);
+  return time.seconds();
+}
+
+template <>
+weekday getFieldType(local_seconds time_point) {
+  auto dp = floor<days>(time_point);
+  return weekday(dp);
+}
+
+bool isWeekday(year_month_day date) {
+  weekday date_weekday = weekday(local_days(date));
+  return date_weekday != Saturday && date_weekday != Sunday;
+}
+
+template <typename FieldType>
+class SingleValueField : public CronField {
+ public:
+  explicit SingleValueField(FieldType value) : value_(value) {}
+
+  [[nodiscard]] bool matches(local_seconds time_point) const override {
+    return value_ == getFieldType<FieldType>(time_point);
+  }
+
+ private:
+  FieldType value_;
+};
+
+class NotCheckedField : public CronField {
+ public:
+  NotCheckedField() = default;
+
+  [[nodiscard]] bool matches(local_seconds) const override { return true; }
+};
+
+class AllValuesField : public CronField {
+ public:
+  AllValuesField() = default;
+
+  [[nodiscard]] bool matches(local_seconds) const override { return true; }
+};
+
+template <typename FieldType>
+class RangeField : public CronField {
+ public:
+  explicit RangeField(FieldType lower_bound, FieldType upper_bound)
+      : lower_bound_(std::move(lower_bound)),
+        upper_bound_(std::move(upper_bound)) {
+    if (!(lower_bound_ <= upper_bound_))
+      throw std::out_of_range("lower bound must be smaller or equal to upper bound");
+  }
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    return lower_bound_ <= getFieldType<FieldType>(value) && getFieldType<FieldType>(value) <= upper_bound_;
+  }
+
+ private:
+  FieldType lower_bound_;
+  FieldType upper_bound_;
+};
+
+template <typename FieldType>
+class ListField : public CronField {
+ public:
+  explicit ListField(std::vector<FieldType> valid_values) : valid_values_(std::move(valid_values)) {}
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    return std::find(valid_values_.begin(), valid_values_.end(), getFieldType<FieldType>(value)) != valid_values_.end();
+  }
+
+ private:
+  std::vector<FieldType> valid_values_;
+};
+
+template <typename FieldType>
+class IncrementField : public CronField {
+ public:
+  IncrementField(FieldType start, int increment) : start_(start), increment_(increment) {}
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    return (getFieldType<FieldType>(value) - start_).count() % increment_ == 0;
+  }
+
+ private:
+  FieldType start_;
+  int increment_;
+};
+
+class LastNthDayInMonthField : public CronField {
+ public:
+  explicit LastNthDayInMonthField(days offset) : offset_(offset) {}
+
+  [[nodiscard]] bool matches(local_seconds tp) const override {
+    year_month_day date(floor<days>(tp));
+    auto last_day = date.year()/date.month()/last;
+    auto target_date = local_days(last_day)-offset_;
+    return local_days(date) == target_date;
```

Comment:
can this underflow, roll over into the previous month? (and is that allowed if so)

Reply:
I've added tests to check this, because I wasnt sure. https://github.com/apache/nifi-minifi-cpp/pull/1335/commits/a7db03ebb2a795bff86aff210e4e29bcf424c372#diff-99182a0f9eeb09751e25ef7c3291f861bbcafcdfb35bfd7adfb666d075aff57eR404-R411
But yeah this can underflow, but it wouldnt match because the date which is checked againts the target date is only incremented.